pkg/docker/config/config.go

```diff
@@ -0,0 +1,271 @@
+package config
+
+import (
+	"encoding/base64"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/containers/image/types"
+	helperclient "github.com/docker/docker-credential-helpers/client"
+	"github.com/docker/docker/pkg/homedir"
+	"github.com/pkg/errors"
+)
+
+type dockerAuthConfig struct {
+	Auth string `json:"auth,omitempty"`
+}
+
+type dockerConfigFile struct {
+	AuthConfigs map[string]dockerAuthConfig `json:"auths"`
+	CredHelpers map[string]string           `json:"credHelpers,omitempty"`
+}
+
+const (
+	authCfg           = ".config/containers"
+	authCfgFileName   = "auth.json"
+	dockerCfg         = ".docker"
+	dockerCfgFileName = "config.json"
+)
+
+var (
+	// ErrNotLoggedIn is returned for users not logged into a registry
+	// that they are trying to logout of
+	ErrNotLoggedIn = errors.Errorf("not logged in")
+)
+
+// SetAuthentication stores the base64 encoding of username and password
+// in the auth file if the credentials are authenticated
+func SetAuthentication(username, password, registry string) error {
+	path, err := getPathToAuth()
+	if err != nil {
+		return errors.Wrapf(err, "error getting path to auth file")
+	}
+	encodedCreds := encodeUserAndPass(username, password)
+	return addAuthToJSON(registry, encodedCreds, path)
+}
+
+// GetAuthentication returns the value of the registry url stored in
+// the auth.json file
+func GetAuthentication(ctx *types.SystemContext, registry string) (string, string, error) {
+	if ctx != nil && ctx.DockerAuthConfig != nil {
+		return ctx.DockerAuthConfig.Username, ctx.DockerAuthConfig.Password, nil
+	}
+
+	// Check containers-storage auth.json file first
+	path, err := getPathToAuth()
+	if err != nil {
+		return "", "", errors.Wrapf(err, "error getting path to auth file")
+	}
+	auth, err := findAuthentication(registry, path)
+	if err != nil {
+		return "", "", err
+	}
+	if auth != "" {
+		return decodeDockerAuth(auth)
+	}
+
+	// check .docker/config.json as secondary option
+	path = filepath.Join(homedir.Get(), dockerCfg, dockerCfgFileName)
+	auth, err = findAuthentication(registry, path)
+	if err != nil {
+		return "", "", err
+	}
+	if auth != "" {
+		decodeDockerAuth(auth)
+	}
+
+	return "", "", nil
+}
+
+// RemoveAuthentication deletes the credentials stored in auth.json
+func RemoveAuthentication(registry string) (string, error) {
+	path, err := getPathToAuth()
+	if err != nil {
+		return "", errors.Wrapf(err, "error getting path to auth file")
```

Comment:
The only way this can currently fail is “error getting path to auth file: error creating directory $dir”, which is contradictory. The `errors.Wrapf` could probably be dropped.

Or, maybe better, split the `getPathToAuth` function into two, one to only get the path (which can’t fail), one to get the path and ensure that the parent exists (which can). That would also mean that read-only users of the config file wouldn't  try to create a directory.

Reply:
fixed.

Reply:
There is one more site, in `GetAuthentication`, — though that error case will probably entirely go away if file creation happens only in the write paths.